### ISSUE TITLE
Potential Memory Leak Graal Native Image in Lambda

### ISSUE THREAD

**Issue by pfrank13** [micronaut-projects/micronaut-aws#385]
### Steps to Reproduce

1. git clone --branch memory_leak https://github.com/pfrank13/java-serverless.git
2. `./deploy.sh` creates `./build/function.zip`
3. deploy that into a lambda (256 MB needed ) that can talk to a mysql instance and execute `POST /books` with `{"name":"someBookName0"}`

This will work about 3 times before an OutOfMemoryException occurs. See attached Cloudwatch logs
[log-events-viewer-result.csv.zip](https://github.com/micronaut-projects/micronaut-core/files/4841603/log-events-viewer-result.csv.zip)

### Expected Behaviour

I would expect the memory footprint wouldn't just continuously go up.

### Actual Behaviour

After about 3 executions you'll get a OutOfMemoryException

```
timestamp	message
1593305669132	START RequestId: 90053c90-1aa6-4e2a-b5c8-a5f56f704471 Version: $LATEST
1593305669140	[36m00:54:29.121[0;39m [1;30m[main][0;39m [34mINFO [0;39m [35mi.m.f.a.p.AbstractLambdaContainerHandler[0;39m - Starting Lambda Container Handler
1593305669288	[36m00:54:29.139[0;39m [1;30m[main][0;39m [34mINFO [0;39m [35mi.m.context.env.DefaultEnvironment[0;39m - Established active environments: [ec2, cloud, lambda, function]
1593305669455	[36m00:54:29.288[0;39m [1;30m[main][0;39m [34mINFO [0;39m [35mcom.zaxxer.hikari.HikariDataSource[0;39m - HikariPool-1 - Starting...
1593305669455	[36m00:54:29.455[0;39m [1;30m[main][0;39m [34mINFO [0;39m [35mcom.zaxxer.hikari.HikariDataSource[0;39m - HikariPool-1 - Start completed.
1593305670020	END RequestId: 90053c90-1aa6-4e2a-b5c8-a5f56f704471
1593305670020	REPORT RequestId: 90053c90-1aa6-4e2a-b5c8-a5f56f704471	Duration: 450.03 ms	Billed Duration: 1200 ms	Memory Size: 256 MB	Max Memory Used: 183 MB	Init Duration: 740.90 ms
1593305677378	START RequestId: 0a2965b4-1c05-47e3-9421-058be78c46e1 Version: $LATEST
1593305677943	END RequestId: 0a2965b4-1c05-47e3-9421-058be78c46e1
1593305677943	REPORT RequestId: 0a2965b4-1c05-47e3-9421-058be78c46e1	Duration: 561.29 ms	Billed Duration: 600 ms	Memory Size: 256 MB	Max Memory Used: 219 MB
1593305680899	START RequestId: 612d6267-7545-4901-84f5-6e0c582d23a4 Version: $LATEST
1593305681243	END RequestId: 612d6267-7545-4901-84f5-6e0c582d23a4
1593305681243	REPORT RequestId: 612d6267-7545-4901-84f5-6e0c582d23a4	Duration: 341.40 ms	Billed Duration: 400 ms	Memory Size: 256 MB	Max Memory Used: 250 MB
1593305684778	START RequestId: d089e2d9-fe82-4880-9968-79ef700c6558 Version: $LATEST
1593305685412	Invocation with requestId [d089e2d9-fe82-4880-9968-79ef700c6558] failed: java.lang.OutOfMemoryError: Direct buffer memoryio.netty.handler.codec.EncoderException: java.lang.OutOfMemoryError: Direct buffer memory
```

### Environment Information

- **Operating System**: Custom Runtime provided by micronaut
- **Micronaut Version:** 2.0.0
- **JDK Version:** N/A Graal Native but based off of 11

### Example Application

```shell
git clone --branch memory_leak https://github.com/pfrank13/java-serverless.git
```

**Comment by graemerocher** [comment]
Thanks for the report, we will investigate. I wonder if it is leaking connections. Could you try something and change 

```
implementation("io.micronaut.sql:micronaut-jdbc-hikari")
```

for

```
implementation("io.micronaut.sql:micronaut-jdbc-tomcat")
```

**Comment by pfrank13** [comment]
Thank you so much for the quick reply! I'll try that tonight and get back to you.

**Comment by pfrank13** [comment]
@graemerocher 
To see what I did it's tag is here:
`git clone --branch 385_memory_leak_tomcat_dbcp https://github.com/pfrank13/java-serverless.git`
In my estimation my config is to only use 1 database connection.

Results ended in an OutOfMemoryException 
[log-events-viewer-result_tomcat_dbcp.csv.zip](https://github.com/micronaut-projects/micronaut-aws/files/4849330/log-events-viewer-result_tomcat_dbcp.csv.zip)

NOTE: I was using 512 MB of RAM because of something I was working on previously but in my testing I think the Graal image is capped at 256 MB, I'm not setting that anywhere so I imagine that's the default? I assume that's just noise to what we are checking.
Important bits
```
timestamp | message
1593488415135 | START RequestId: 9d9b4b1b-014b-4c70-b9ab-41fe6000979c Version: $LATEST
1593488415143 | [36m03:40:15.125[0;39m [1;30m[main][0;39m [34mINFO [0;39m [35mi.m.f.a.p.AbstractLambdaContainerHandler[0;39m - Starting Lambda Container Handler
1593488415143 | [36m03:40:15.143[0;39m [1;30m[main][0;39m [34mINFO [0;39m [35mi.m.context.env.DefaultEnvironment[0;39m - Established active environments: [ec2, cloud, lambda, function]
1593488415949 | END RequestId: 9d9b4b1b-014b-4c70-b9ab-41fe6000979c
1593488415949 | REPORT RequestId: 9d9b4b1b-014b-4c70-b9ab-41fe6000979c	Duration: 531.44 ms	Billed Duration: 1100 ms	Memory Size: 512 MB	Max Memory Used: 184 MB	Init Duration: 535.22 ms
1593488422192 | START RequestId: d1cfb777-4ead-4234-a2d9-ac09779f553f Version: $LATEST
1593488422427 | END RequestId: d1cfb777-4ead-4234-a2d9-ac09779f553f
1593488422427 | REPORT RequestId: d1cfb777-4ead-4234-a2d9-ac09779f553f	Duration: 232.51 ms	Billed Duration: 300 ms	Memory Size: 512 MB	Max Memory Used: 218 MB
1593488427108 | START RequestId: ad0c42c8-a6c9-4804-8f82-56554f5bea6d Version: $LATEST
1593488427246 | END RequestId: ad0c42c8-a6c9-4804-8f82-56554f5bea6d
1593488427246 | REPORT RequestId: ad0c42c8-a6c9-4804-8f82-56554f5bea6d	Duration: 134.55 ms	Billed Duration: 200 ms	Memory Size: 512 MB	Max Memory Used: 251 MB
1593488431789 | START RequestId: dee7f9a1-76d1-49d5-a569-dcf45dc7761b Version: $LATEST
1593488432376 | Invocation with requestId [dee7f9a1-76d1-49d5-a569-dcf45dc7761b] failed: java.lang.OutOfMemoryError: Direct buffer memoryio.netty.handler.codec.EncoderException: java.lang.OutOfMemoryError: Direct buffer memory
1593488436549 | END RequestId: dee7f9a1-76d1-49d5-a569-dcf45dc7761b
1593488436549 | REPORT RequestId: dee7f9a1-76d1-49d5-a569-dcf45dc7761b	Duration: 4757.88 ms	Billed Duration: 4800 ms	Memory Size: 512 MB	Max Memory Used: 257 MB
1593488436549 | RequestId: dee7f9a1-76d1-49d5-a569-dcf45dc7761b Error: Runtime exited without providing a reason Runtime.ExitError
```

**Comment by gradinac** [comment]
Hey @pfrank13, thank you for reporting the issue!
I have been looking into this from the native-image side. I do not have access to AWS so I could not run the lambda version of this project.

When running the application locally (without the Lambda stuff in it), I have not been able to hit the error. 
One of the main leads here is the exception you've posted: `java.lang.OutOfMemoryError: Direct buffer memory`. This seems to indicate that the allocation problem does not lie in allocating Java objects but rather the allocation of direct buffers.
One of the things I've noticed is that, despite the heap used by the program being relatively small, we get a huge amount of data in the off-heap section. The native-image RSS (Resident Set Size) capped at 700MB, but I've seen the Java version of it go even higher.

Have you tried running the Java version of this lambda in the exact same setting? What were the results?

Is the `Max Memory used` the total physical memory used by the application? By default, `native-image` will cap the heap size to 80% of the physical memory on the system. You can override this by passing `-R:MaxHeapSize=<val>`, where `<val>` is the new heap size in bytes. You should pass this option to the built image (e.g. in the bootstrap script), not to the image builder.

If possible, could you try running the native-image in a beefier container, perhaps one with 1GB of RAM? The way I emulated load locally was with the poor man's request spammer: `watch -n0.2 'curl localhost:8080/books/<isbn>'`. This is what caused the RSS to spike up.

I will do some more digging on this in the meantime - if I come across anything new, I will let you know.

**Comment by pfrank13** [comment]
Sure thing! The Java version of the Lambda had no issues at all. Worked like a charm and memory was stable (albeit ~380 MB which isn't that surprising). So this is definitely something with the Native portion. `-R:MaxHeapSize=<val>` yeah for whatever reason in my 512 MB lambda it was capping (it seemed) at 256 MB but I'll try that and get back to you ASAP

**Comment by pfrank13** [comment]
All right, so hopefully you can explain this. Since you had me monkeying around in the bootstrap I was looking at the params being passed and they were (as generated by the micronaut starter site) `./java-serverless -Xmx128m -Djava.library.path=$(pwd)`. I figured, well if the starter generated it, then I guess it's fine. I added the -R argument you said, no dice, still OOME. However, for fun I just removed the Java related cruft (as I can't imagine it was being used). Problem goes away entirely....  So if you can explain that to me I'm all ears :) 
Also, I don't know that the "fix" is here at that point, almost seems like the starter needs to be updated or at least maybe have a warning somewhere?

**Comment by pfrank13** [comment]
Sorry I was speaking in Spring-isms. I guess for micronaut then starter site is called launch. Every time you see "starter" above I meant this site: https://micronaut.io/launch/

**Comment by gradinac** [comment]
@pfrank13 That argument completely slipped past my mind! `native-image` respects the `-Xmx` option which limits our available memory at runtime. It is directly tied to `MaxHeapSize`, but it might be that, due to the ordering of the arguments, the generated one overrode the one you added. I'll see if we can handle that case on `native-image` side with some kind of a message informing the user that they passed two different options that do the same thing.
The reason I did not hit the exception was because I generated and ran the `native-image` by hand.

**Comment by gradinac** [comment]
@pfrank13 Nice work catching that! :)

**Comment by pfrank13** [comment]
Mystery solved. Works as designed! Thank you so much for your help @gradinac, at least we found it quickly and I didn't make you rabbit hole too bad on a non-issue :)

**Comment by graemerocher** [comment]
@pfrank13 could you specify what is the argument that is problematic?

**Comment by graemerocher** [comment]
We may need docs

**Comment by gradinac** [comment]
@pfrank13 No problem at all! I'm glad we got to the bottom of it :) 

@graemerocher The problematic argument was '-Xmx128m', it limited the amount of memory available to the app leading to the OOM exception

**Comment by pfrank13** [comment]
Yeah +1 to @gradinac , I think that may have done it. I'm fairly certain that is auto generated by the launcher site cause I'm not smart enough to know that :)

**Comment by ilopmar** [comment]
@pfrank13 Yes, that value is generated by the Starter: https://github.com/micronaut-projects/micronaut-starter/blob/master/starter-core/src/main/java/io/micronaut/starter/feature/awslambdacustomruntime/templates/bootstrap.rocker.raw#L15

Please feel free to send a PR to remove it or to increase it.

I also think we should remove the `-Djava.library.path=$(pwd)`. That was used in the past when we needed to add the `libsunec.so` library, but that is not necessary anymore.
